Added support for loneoceans Lume1 Lume X1 designs and some Anduril updates

loneoceans opened the pull request:
Added support for the Lume X1 driver design, and added some additional functionality in Anduril2. 

_Changelist:_

* Updated MODELS and BRANDS with loneoceans specific designs
* Added Lume 1 and Lume X1 definitions
* Added Beacontower mode in blinky modes
  * Requires defining USE_BEACONTOWER_MODE to enable
  * Changes also made for each blinky mode to cycle through
  * Commit 26 Dec with some bugfixes
* Added support for RGB aux LEDs on different ports
  * Requires defining AUXLED_RGB_DIFFERENT_PORTS
* Added support for long blink for negative sign during temp check
  * Requires defining USE_LONG_BLINK_FOR_NEGATIVE_SIGN
* Added loneoceans Lume-X1-40W definitions including arch, hwdefs, config, and helper functions
* Added loneoceans Lume1-6AF definitions including arch, hwdefs, config, and helper functions
  * Commit on 26 Dec with some ramp handling bugfixes for Lume-X1

Lume-x1-40w and lume1-6af binaries tested working on Lume X1 / Lume1 development platforms. RGB LEDs working on different ports, as well as Beacontower mode. Temperature check (negative sign) tested working by placing driver inside freezer. Added functionality makes use of ifdefs since smaller MCUs like Attiny85 do not have sufficient flash. All 74 hex builds succeeded using make command with no errors.

loneoceans committed:
updated models and brands with loneoceans specific designs

loneoceans committed:
added beacontower mode, long blink for neg sign in tempcheck, support for rgb aux on different ports

loneoceans committed:
initial add of loneoceans lume x1 definitions including arch, hwdefs, config, and helper functions

SammysHP commented:
Thanks for your contribution and updating the patches for your great drivers! You might want to check the code for the FW3X-lume1: https://github.com/ToyKeeper/anduril/tree/trunk/hw/lumintop/fw3x-lume1

loneoceans committed:
small bug fixes for beacontower mode and improved comments

loneoceans committed:
bug fix for lume-x1-40w ramp control handling

loneoceans committed:
add support for lume1-6af driver

loneoceans commented:
Added support for the Lume1-6AF driver, as well as some bugfixes found for Lume X1 driver and Beacontower mode. TODO: create definition files for relevant flashlights which use the Lume drivers, such as some FireflyLite flashlights, as well as write more detailed documentation for these drivers.

loneoceans committed:
small adjustments for lume1-fw3x driver

loneoceans committed:
added support for turbo temperature boost as an option

loneoceans committed:
added hw-config files for some fireflylite flashlights, updated MODELS and BRANDS lists, and added some documentation

loneoceans commented:
Added additional config files for some FireflyLite flashlights (those that use the Lume1 and Lume X1 drivers), as well as took a stab at the E12C (needs testing). For E12C, had some trouble trying to fit firmware into Attiny85 due to flash constraints with 3 PWM tables, and had to undef sunset timer (uses almost 500 bytes) to compile. Updated MODELS and BRANDS list. Todo: add config for older 1634 Lume1 drivers.

loneoceans committed:
added support for the older 2020 fireflies lume1 driver (attiny1634)

loneoceans commented:
Added config files for the older Attiny1634 Lume1 drivers.

loneoceans committed:
improved fireflies-e12c ramp table, added sunset timer but removed momentary and sos modes

loneoceans committed:
fixed error in pin toggle

Light-Veteran commented:
Thank you for all your support and for your awesome drivers

SiteRelEnby commented:
Tested 6af today, looks good, everything working as it should. Will work on the others as time allows.

Edit: It does seem to be getting hotter than it should. 67 degrees on tempcheck and 60 degrees externally using IR, from a turbo run on stock thermal settings. I'll have a look as well.

SiteRelEnby commented:
@loneoceans if you don't mind answering a couple of questions:

Edit: see below,  some of these don't matter now.

* The 6af model code has changed from the stock firmware shipped on the T1R. Do you happen to know if the T1R uses FET/noFET? I've been testing with noFET on an FFL505A T1R so far but just realised that I have no idea if that emitter can take FET or not. (Of course, like an idiot, I didn't backup the original firmware on my development T1R before reflashing it...)
~~* Related: Assuming the SFT40 uses FET and the W2.1 noFET? (Edit: looks like SFT40 and W2.1 are both noFET I guess (by process of elimination as I know the W2.1 can't), same model code + same checksum when I dumped the firmware)~~
* What is the design goal of the thermal config in this PR? I measured my T1R reaching as high as 62 degrees externally and 67 on the ADC. Are you trying to increase the turbo hold time more, or reduce the magnitude of a hard drop? The two are somewhat mutually exclusive, and as it is, with no configuration or changes, my test T1R gets hot enough to hurt even my hand, and I have a higher hot light tolerance than most.
~~* What are your thoughts about not using `TURBO_TEMP_EXTRA` on noFET, or setting it to a smaller value? When I hear back about if FET is safe to use with the FFL505A I'll test the FET version as well but don't want to accidentally kill my dev light. I can't definitely test it yet until the above are answered but I feel like the lack of magnitude of stepdown from noFET may not be the same problem in the FET versions (I have a lume1 FW3X as well as the 2020 and X1 lume1 lights and neither of those stay as hot as this T1R does).~~

The problem righ now that when I tested it, it would never step down from turbo, and exceed the temp limit even with `TURBO_TEMP_EXTRA` considered - if `TURBO_TEMP_EXTRA` is going to be used then the thermal regulation under that point should probably be made more aggressive.

SiteRelEnby commented:
...and I just realised that the fireflies-specific config overrides the default thermal limit to 60, *that's* why no matter what I changed I couldn't get it under there, heh... IMO that's too high for such a small light, *especially* with `TURBO_TEMP_EXTRA`.

Comment added: https://github.com/ToyKeeper/anduril/pull/37/files#r1547006604 - with a value of 50 here, 6af looks good.

SiteRelEnby commented on hw/fireflies/lume1-6af/anduril.h:
```diff
@@ -0,0 +1,17 @@
+#include "loneoceans/lume1-6af/anduril.h"
+
+#undef DEFAULT_THERM_CEIL
+#define DEFAULT_THERM_CEIL 60
```
60 here results in temperatures externally of over 60, and internal ADC temperatures close to 70, which is dangerous for the battery. 50 seems like a good value here, external temp never went over 55 and ADC 60.

SiteRelEnby commented:
Going to work on testing X1 next, just need to collect some data with the stock firmware first, should be able to within the next couple of days unless something comes up.

loneoceans commented:
Hi @SiteRelEnby, so glad to have you help review the code! I really appreciate it, along with the many other contributions you have made. Thank you.

Answering your questions, for the T1R, as far as I am aware, none the emitters that Jack paired with it should be driven via FET. The hardware actually supports PWM though, so you could set your own reduced-power Turbo mode. But in general I am not quite fan of unregulated FET drive. I believe the T9R had a SBT90 option, which can take the full 100% duty cycle FET. As a side note, I have an updated driver topology being finalized which will mitigate these issues. 

For thermal settings, I absolutely agree with you, and my preferred default thermal threshold is 50C; this is why they are set as such in the loneoceans HW definitions. However, I am sure that you are aware of the desire from perhaps a small but vocal group of hobbyists about sustained brightness and turbo modes, which have led to the request from FFL to set the default higher. I tried to push for and succeeded with a default of 50C for the newer Lume X1 boost drivers, but let me try to reach out to Jack again about the defaults. 

In addition, the therm_warning thresholds and response magnitudes definitely can use more fine-tuning! I'm not quite sure if I fully understand how these variables are actually handled, or their design intent, so I'd love to get your help on these values.

Thanks again for offering to help out with the X1, if you need any additional hardware or such, please let me know. I did message you on Reddit previously but I'm not sure if you've seen it, but I'll be happy to answer any questions over chat or another more convenient platform if you'd like.

SiteRelEnby commented:
@loneoceans I tested a few different builds with different thermal config before I noticed the fireflies hwdef was overriding the thermal limit and they didn't make that much difference when the limit was raised (i.e. having the limit at 60 made way more of a difference than when I was finetuning the thermal regulation even to be significantly more aggressive), although I didn't test as many combinations as I would want to be 100% certain yet, and I'd say that most of them are reasonable for a high efficiency driver,

I'm well aware of people wanting a hotrod, I'm one of the more vocal hotrod lovers after all 😜, but the highest ADC temperature I *observed* was 63 degrees externally via IR, and ADC temp of 68, which is reaching the point of being dangerous to the battery - perhaps not with a Molicel, but it's more than I'd want to push something more marginal.

The issue was that when the light heatsoaked, it took a long time for thermal regulation to drop it about 4-5 levels (if I had to estimate, I didn't think to set mode memory up before trying that run...).

I think that maybe with emitters that use FET it's fine (my E07X gets just as hot but actually drops off much faster in heat generation when it throttles down out of turbo, while the T1R stays heatsoaked just *barely* under its thermal max for way longer) as with FET there's a bigger drop off in the rate of heating from only a single level of stepdown taking away a good 25-50% of the actual output when only dropping a single level, but with regulated only like the T1R, the main issue I ran into was that it was stepping down slowly enough that at level 145 or so, the buck driver's output was still enough to cause the light to overheat. In addition, I found that even when deliberately heatsoaking the light, I noticed that sometimes the ADC would not *quite* read high enough to cause even a single stepdown for too long to be subjecting a battery to.

My suggestion would be to either make thermal regulation somewhat more aggressive in general on noFET (145ish is not an appropriate level to step down to after 5+ mins of heat soaking on a light where the power curve by ramp level is largely flat, as opposed to a big kick up at the end with a DD FET), and/or to reduce `TURBO_TEMP_EXTRA` to 5 on noFET. I can test a few combinations tomorrow and see what works, I have a lumen tube and IR camera to test with, and I'm happy to push output as high as it's safe to do so.

SiteRelEnby commented:
For FET, it's probably fine as it is, there's a big output drop in just the one level.

For noFET, with an effective limit of 70 in turbo, it was reaching 63 externally, and was holding that temperature without stepping down at all (i.e. when 2Cing the light, it dropped back to the original level as opposed to going back up to turbo). Presumably as it has not stepped down yet that correlated to 69 internally.

For noFET, sustained upper ramp output is always going to be fairly good, and since there's no real perceptible difference between turbo and 149, I was thinking perhaps a thermal limit of either 50 +10 for turbo or 45+15. That way, the light holds turbo for as long as possible, but just has a larger stepdown to a safe sustainable thermal level - I tested with 50 + 10 and the maximum external temp was 55 - if you think 60 in general is acceptable as a maximum external temp it *can* reach, would 50+15 or even 55+10 be acceptable? That way it *can* reach 60, but won't remain at battery-damaging temps internally.

I have no problem with the light being generally set up to run hot, but it can't be to the level that it risks burning someone or causing damage to the battery (60 degrees is only safe for skin contact for 5 seconds - I very lightly burnt my finger trying to get it into tempcheck after one test run) - the higher limits might be absolutely be fine in a larger light but the driver is efficient enough that even with the 22430 tube the T1R gets dangerously hot. It's definitely an excellent driver, the main reason it's a problem is how well it does hold turbo without dropping due to battery voltage sag after all - I'm learning driver development at the moment and I completely get wanting to push the limits, and I will point out that even with thermal regulation turned up a bit it still outperforms a T1R with original firmware ;)

If you have alternative solutions I'd be happy to test them, but in general, I feel like the goal should be to keep the real temperature below 60 as soon as thermal regulation *has* started, but think that pushing it hast 60 briefly for the sake of the initial turbo is probably less of a problem.

Another suggestion, probably not enough to *entirely* mitigate the problem on its own but something else to consider if pushing the limits on thermals is to set simple UI's 2C style to anduril 2 combined with a ceiling of 130 or so, so the user needs to ramp all the way up to get 150, while you can leave the full ramp unlocked with instant turbo from anywhere in advanced.

re. Reddit, I've decided to make an effort to check Reddit Chat more often, I've missed various stuff before, it's just not that convenient to check as I use old.reddit.com but I'll put some time into finding a solution. I'm probably done for this evening unless I can't sleep, but I'll look at this more tomorrow.

loneoceans commented:
Thanks again SiteRelEnby for helping take a deeper look into this, especially with all the testing with different settings and for your very helpful suggestions!  

For the noFET config, the idea was to keep the same ramp table, just without the FET channel. The RAMP_SIZE is redefined as 149, and the ramp tables are as well, removing the last value. In addition, the TURBO_TEMP_EXTRA was only intended to be used for flashlights with FET enabled. It should be #undef-ed in the noFET headers, thus the default temperature handling in fsm/adc.c should take over. It sounds like to me that maybe the flashlight is still getting into the TURBO_TEMP_EXTRA mode (did I use the undef directive incorrectly?) That could explain why it was hitting such high temperatures. Alternatively, redefining TURBO_TEMP_EXTRA as 0 should effectively default to the regular temperature stepdown handling as well. 

As for your suggestions, I think it may also make sense to set the default ceilings on the noFET version to be lower than the full 6A which it is currently set to. One thing I had trouble with was configuring the temperature response parameters because the driver is used in a few different flashlights (the original intent was for it to be in the E12R only, but Jack decided to use it many other flashlights with vastly different sizes and thermal capabilities). Another way we could go about doing this is to define THERM_RESPONSE_MAGNITUDE to a larger value such as 64 for the noFET version, which should increase the stepdown aggressiveness. I'm going to try to do some additional testing as well on my side, and once we finalize on a good set of values, let's adjust the defaults.

SiteRelEnby commented:
I actually missed that it did get undef'd in noFET (or otherwise I was reading some test-hotpatched code I forgot to revert, I guess). I'll test with some increased thermal regulation, I guess the way it was sitting at at 68-69 specifically made me wonder if the +10 was the cause. In terms of improving sustained performance, reducing the lookahead a bit is probably similar in effect but with less overheating possible.

SiteRelEnby commented:
PR submitted: https://github.com/loneoceans/anduril/pull/1

With these changes, PR looks good and ready to merge when that one is.

loneoceans committed:
update temperature thresholds to lower defaults for lume1-6af drivers, validated in several flashlights

loneoceans commented:
> PR submitted: [loneoceans#1](https://github.com/loneoceans/anduril/pull/1)
> 
> With these changes, PR looks good and ready to merge when that one is.

Thanks @SiteRelEnby for the feedback and suggestions. Did more testing since I finally got more flashlights on hand, validated with FLIR thermal camera. Requested FireflyLite to reduce thermal defaults. Found a bug in level indexing. Also adjusted default ceiling for the boost driver to be slightly lower, for better sustained performance and more obvious turbo mode.

loneoceans committed:
update ceiling defaults and improvements in ramp tables for lume-x1-40w

loneoceans committed:
added support for lume x1 driver with avr32dd20

ToyKeeper commented:
I'm going to try to merge this soon, but it's not trivial and I'll need to change some things along the way.  Some questions and thoughts:

Models:

- Could you add some `README.md` files for any drivers which aren't named after a model of flashlight, to give more information about the driver and what products it is used in, and generally provide any relevant info or links for people to learn more?

- Are some of these driver models duplicates?  For example, it's unclear what is the difference between a `fireflies/lume1-6af` and a `loneoceans/lume1-6af`, or which products it is used in, or if both are needed.  I'd like to avoid duplicates and prototypes.  It looks like most of the `loneoceans/` models are duplicates or prototypes for other models, like `loneoceans/lume1-2020` may not be needed because `lumintop/fw3x-lume1` already exists.

- If a particular driver is used in only one light, or in only one brand, it should probably use a model number with that brand's prefix.  I've also been avoiding non-leading zeroes in model numbers, to reduce confusion between 0 and 1 blinks.  Mind if I change some model numbers?  It would help to know which product(s) use each build.

Ramps:

- I'd like to remove TURBO_TEMP_EXTRA if possible, and use a smooth ramp up to full power instead of having a big jump from level 149 to 150.  Without a big jump, it probably won't need an "extra" turbo, and thermal regulation will be less likely to overshoot.  I have a FW3X to use for testing this, but I don't have any other models with a big 149-to-150 jump.

- I'd like to recalculate some ramp tables to keep the Vref from changing back and forth.  When it goes "V05, V05, V25, V11, V11, V15, V25, V15, V15, V05", that breaks gradual adjustments and interferes with thermal regulation and sunset mode.  It also increases the chance of a zig-zaggy ramp if the reference voltages are slightly different than expected.  It should try to minimize the number of Vref changes if possible, ideally only one time per "gear" or "path" as it goes up the ramp.  I don't have hardware for most of these drivers though, so I can't really calibrate it myself.

- I have a light which identifies itself as a lume-x1-avr32dd20 that I plan to use for testing and calibration.  However, it does not emit any light at the lowest few ramp levels, and the ramp is bumpy (perhaps because of the zig-zag Vref values mentioned above).  I may need help testing changes to the ramp, to ensure it works on more than one item.

- Ramp size 156.  Why?

Misc:

- RGB on separate ports is definitely needed, but I might try to do it in a more generic way.  I didn't properly abstract it in the last rewrite, and it has been bugging me ever since.  But now that there are lights which actually need it, I should probably stop avoiding it and get it done.

- Beacon Tower mode seems like it could probably use the same animations as smooth steps, and perhaps could be an option in beacon mode instead of being a separate mode.  Does this need to be a separate mode, or would it be okay to merge it into the regular beacon mode?

- I might try to merge the hardware support code by itself, and merge UI features separately.  These sorts of things are usually better handled as separate patches.

Anyway, we should be able to get these merged... but it might take a little while to clean up the changes first.

loneoceans commented:
Hi ToyKeeper, very good to hear from you, and thanks so much for the thoughtful comments. Responding to your points:

**Models:**

- Yes I can certainly do so; I actually did write up a Readme.md file in the hw/fireflies directory in this commit: https://github.com/ToyKeeper/anduril/pull/37/commits/897364a2a071e3e59930960f1d563f6f3d9e0d75. Do you think that is sufficient? I also made some adjustments to the Readme you wrote (thank you) for the Lume1-FW3X.

- About hw/loneoceans and the hw/fireflies, understand your point about prototypes etc. I was thinking that the hw/loneoceans definition would serve as the base-definition for Lume1, Lume X1 etc drivers, and that the company-specific definitions would use the loneoceans definitions, along with with company-specific tweaks such as UI configs, ramp tables, temperature settings etc. That seems to follow the same scheme as some of the other exisiting lights. Would be happy to hear your thoughts on how best to rearrange this. 

- Regarding a driver being used only in one light / brand, will be happy to discuss with you separately, but otherwise totally OK with you reorganizing the numbers to reduce confusion. 

**Ramp:**

- TURBO_TEMP_EXTRA - Agreed here, I am OK removing this, but would like to hear your thoughts on how best to implement the thermal throttling for those drivers (specifically the Lume1 6A+FET designs… I have some newer designs which avoid needing it). 

- Agreed with the point about zigzagging between different Vrefs. This is much less of an issue with the AVR32DD20 with the higher resolution DAC; I can definitely generate ramp tables which don’t bounce back and forth (though having tried both, it does give a smoother ramp esp. for the Attiny1616).

- Note that I haven’t had any issues with the Vref accuracy (+-3%), or DAC output, for that matter. For Microchip to guarantee 8bit or 10bit DAC operation, the internal Vrefs and DAC resolution must meet the required specification. In the (admittedly small handful) of samples I’ve seen, they all seem to meet that. I verify output current of my drivers with a dedicated current probe and see good ramp characteristics for all individual levels. Regardless, point taken about gradual transitions which is most obvious in sunset mode for example. 

- Ramp Size 156 for Lume X1 AVR32DD20 – the reasoning was to add a few steps because the lowest moonlight mode is very low (some of the lower modes looks like it’s off but it is, in fact, on, may require camera to see, agreed that it’s impractical), therefore I thought I’d add 5 more steps to the ramp, since most people default to somewhere between level 5 and 10 for their floor and some were bummed about losing a few levels. When refining the ramp table (via actual measurement), I added one more level by hand to smoothen out a discontinuity, but this table is very much preliminary since I don’t have production hardware for it yet*.

- For hardware, been wanted to get some hardware over to you, see below. 

**Misc:**

- RGB on separate ports – I have that working on both the 1616 and 32DD20, glad to see that rolled in just for flexibility. Regarding RGB, would also like to discuss some additional items with you. 

- I added Beacontower as a for-fun mode when a community member requested it many years back. Back then (before smooth ramping), it seemed like a fun effect. Would be happy to remove it or roll an additional feature into the regular beacon mode to allow for a smooth ramp up and down. Admittedly a low priority feature. 

- Understand your point about hardware support and UI features. 

As a side note, would appreciate any feedback on how I should organize my firmware management / best practices to allow for cleaner pull requests in the future. 

Regarding hardware – may I know what the best way to reach you is? I dropped a few chats/PMs on Reddit, as well as sent you some emails (not sure which is best for you?) over the past few months but not sure if you had received them. 

My email, [loneoceans@gmail.com](mailto:loneoceans@gmail.com) is the best way to contact me. Have a bunch of hardware I’d like to get to you, as well as to discuss some *upcoming developments. 

As always, greatly appreciate your time and contributions to the community. Thank you.

loneoceans committed:
updated ramp tables for 150 levels, adjusted Vref table as per ToyKeeper's request

loneoceans committed:
updated ramp tables for ATTINY1616 Lume X1 with progressive Vref transitions, adjusted ramp table for AVR32DD20 Lume X1, added pin toggle (for OTG) in momentary mode

loneoceans committed:
adjusted config files for lume-x1-avr32dd20 as well as MODELS

loneoceans committed:
small adjustment to ramp table for lume-x1-avr32dd20

loneoceans committed:
small adjustments to defaults, strobe, blink values

loneoceans committed:
added readme file for the hank-lume-x1 driver

loneoceans committed:
added config files for fireflies-lume-x1-avr32dd20

DurvalMenezes commented on MODELS:
My SBT90A T9R which I just received from FFL, blinks 0451 as model, not 0452 as you indicated.

Is there a possible divergence in model numbers between what you're using here, and FFL's actual practice? 

OTOH I'm pretty sure my T9R has FET as it blinks when ramping up, just before reaching maximum brightness -- so perhaps your model number is correct and consistent with FFL after all.

DurvalMenezes commented on MODELS:
```diff
@@ -55,6 +55,10 @@ Model     MCU         Name
 0441      attiny85    fireflies-e01
+0445      attiny85    fireflies-e12c
+0451      attiny1616  fireflies-lume1-6af       (E07X, E12R, NOV-Mu, etc)
+0452      attiny1616  fireflies-lume1-6af-nofet (T9R, T1R, etc)
```
My SBT90A T9R which I just received from FFL, blinks 0451 as model, not 0452 as you indicated.

Is there a possible divergence in model numbers between what you're using here, and FFL's actual practice?

OTOH I'm pretty sure my T9R has FET as it blinks when ramping up, just before reaching maximum brightness -- so perhaps your model number is correct and consistent with FFL after all.

DurvalMenezes commented on MODELS:
```diff
@@ -55,6 +55,10 @@ Model     MCU         Name
 0441      attiny85    fireflies-e01
+0445      attiny85    fireflies-e12c
+0451      attiny1616  fireflies-lume1-6af       (E07X, E12R, NOV-Mu, etc)
+0452      attiny1616  fireflies-lume1-6af-nofet (T9R, T1R, etc)
```
My SBT90A T9R which I just received from FFL, blinks 0451 as model, not 0452 as you indicated.

Is there a possible divergence in model numbers between what you're using here, and FFL's actual practice?

OTOH I'm pretty sure my T9R has FET as it blinks when ramping up, just before reaching maximum brightness -- so perhaps your model number is correct and consistent with FFL after all.

Light-Veteran replied:
Yes my friend, T9R with SBT90 has FET

loneoceans replied:
@DurvalMenezes, Light-Veteran is correct. The T9R comes in different emitter configurations. Some of them can take a much higher forward current such as your SBT-90, and therefore use the with-FET firmware. Others like the SFT-40 cannot and need to use the no-FET firmware. I had written a readme file in the Fireflies HW folder; hopefully that will help:
https://github.com/loneoceans/anduril/blob/trunk/hw/fireflies/README.md

DurvalMenezes replied:
Thanks for the clarifications, @Light-Veteran and @loneoceans. 

Indeed, I've since compiled the fireflies-lume1-6af HEX and  flashed my SBT90.2 T9R with it, and everything seems good. Consulting the README.md was useful as I was able to match the revision number silk-screened in my T9R driver PCB ("REV C 10-22") with the exact model number it blinks during 15C, that is, 0451.

That said, I think the above-mentioned MODELS and README.md files could be clearer.  I will try and later submit a patch to that effect.

DurvalMenezes replied:
> That said, I think the above-mentioned MODELS and README.md files could be clearer. I will try and later submit a patch to that effect.

@loneoceans I'm working on that. I think that just stating "a single small emitter" as in the currend README.md file is not the best; I think the best would be to indicate an objective number, like maximum forward current, which users can then consult on the datasheet for their particular LED, and decide on whether to flash the 'plain' or the nofet version.

Is it possible to indicate such a number?

loneoceans committed:
added AVRXMEGA3 define for thermal cal reset as suggested by bjrolfe

loneoceans committed:
updated MODELS, added lume1-15af

loneoceans committed:
added lume-x1-avr32dd20-48w variant

GarikFirst commented:
Is there any chance this will be merged?

loneoceans committed:
added config files for fireflies and hank lume-x1-32w variant, updated MODELS

loneoceans committed:
roll in some Anduril updates including POVD changes, vbat precision, button aux threshold setting, docs, etc

loneoceans committed:
move hank-lume-x1-32w into x1 subfolder

loneoceans committed:
small adjustments to lume1 and lume x1 drivers

loneoceans committed:
fix lume1 ADC handling broken in previous commit

loneoceans committed:
add TWI/I2C driver and advanced AUX peripheral support for anduril

loneoceans committed:
add support for lume10 and config files for fireflies-lume10

loneoceans committed:
add config files for lume-x1-c

loneoceans committed:
fix bug in auto-lockout mode rainbow adv-aux handling, and a typo

loneoceans committed:
update config files for lume-x1-c

ToyKeeper commented:
Hi, looking through the patch again, I noticed something which may be a problem.

At first, I thought the extra code in `drivers/` was BSD licensed, which would be fine because BSD 3-clause is GPLv3 compatible.  But this is not a standard BSD 3-clause license.  It includes an extra clause:

> 4. This software may only be redistributed and used in connection with an Atmel micro controller product.

This does not seem to be compatible with the GPLv3 license used by Anduril, because it places corporate restrictions on use and redistribution.  It's a custom license which hasn't been officially evaluated for GPL compatibility, which makes it a risk, but the extra clause seems *very* unlikely to be compatible.

There are two ways to resolve this:

1. Remove the code with the custom license.  Maybe replace it with something using GPLv3 or a compatible license.
2. Transfer copyright from all contributors to a single legal entity, and then grant vendors a non-GPL license to use this code in their products.

Option 1 is far more practical and more aligned with the open nature of the project.

So we're going to need a TWI driver with a better license, or just write one from scratch.

loneoceans committed:
add support for lume-n1, hank lume-n1 variants, updated MODELS
